test/extensions/filters/network/thrift_proxy/conn_manager_test.cc

```diff
@@ -1058,6 +1058,53 @@ TEST_F(ThriftConnectionManagerTest, ResetDownstreamConnection) {
   EXPECT_EQ(0U, stats_.request_active_.value());
 }
 
+TEST_F(ThriftConnectionManagerTest, RequestWithMaxRequestsPerConnection) {
+  const std::string yaml = R"EOF(
+stat_prefix: test
+route_config:
+  name: local_route
+max_requests_per_connection: 1
+)EOF";
+
+  initializeFilter(yaml);
+  EXPECT_EQ(1, config_->maxRequestsPerConnection());
+
+  writeComplexFramedBinaryMessage(buffer_, MessageType::Call, 0x0F);
+
+  ThriftFilters::DecoderFilterCallbacks* callbacks{};
+  EXPECT_CALL(*decoder_filter_, setDecoderFilterCallbacks(_))
+      .WillOnce(
+          Invoke([&](ThriftFilters::DecoderFilterCallbacks& cb) -> void { callbacks = &cb; }));
+
+  EXPECT_EQ(filter_->onData(buffer_, false), Network::FilterStatus::StopIteration);
+  EXPECT_EQ(1U, store_.counter("test.request_call").value());
+
+  writeComplexFramedBinaryMessage(write_buffer_, MessageType::Reply, 0x0F);
+
+  FramedTransportImpl transport;
+  BinaryProtocolImpl proto;
+  callbacks->startUpstreamResponse(transport, proto);
+
+  EXPECT_CALL(filter_callbacks_.connection_.dispatcher_, deferredDelete_(_));
+  // Since max requests per connection is set to 1, the connection will be disconnected after the
+  // first request is completed.
+  EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::FlushWrite));
+
+  EXPECT_EQ(ThriftFilters::ResponseStatus::Complete, callbacks->upstreamData(write_buffer_));
+
+  filter_callbacks_.connection_.dispatcher_.clearDeferredDeleteList();
+
+  EXPECT_EQ(1U, store_.counter("test.request").value());
+  EXPECT_EQ(1U, store_.counter("test.request_call").value());
+  EXPECT_EQ(0U, stats_.request_active_.value());
+  EXPECT_EQ(1U, store_.counter("test.response").value());
+  EXPECT_EQ(1U, store_.counter("test.response_reply").value());
+  EXPECT_EQ(0U, store_.counter("test.response_exception").value());
+  EXPECT_EQ(0U, store_.counter("test.response_invalid_type").value());
+  EXPECT_EQ(1U, store_.counter("test.response_success").value());
+  EXPECT_EQ(0U, store_.counter("test.response_error").value());
+}
```

Comment:
Can we also test:

* limit set but not reached
* limit set, reached, cleared and next request goes through

And also test the suggested stat from above is properly incremented in each case.